### PR TITLE
feat(UIForms): date time

### PR DESCRIPTION
**Pull request by jsomsanith**
**What is the problem this PR is trying to solve?**
No widget for date/time picking

**What is the chosen solution to this problem?**
Component change : 
* InputDateTimePicker accepts timestamp, Date, or string following dateFormat as value

Form : 
* form widget
* iso-datetime and timestamp format with validation
* if string is required, without any format, the uiSchema widget configuration defines the format. It is used for display, and value is passed as is to the form.
* InputDateTimePicker error message is displayed as is for now. Another PR to have a proper translated error system.

**Please check if the PR fulfills these requirements**

* [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
* [x] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/forms/src/UIForm/fields/Date/DateTime/DateTime.component.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import { isoDateTimeRegExp } from '../../../customFormats';
+import { WidgetUnhandleTypeError, WidgetUnexpectedTypeError } from '../WrongTypeError';
+import { generateDescriptionId, generateErrorId } from '../../../Message/generateId';
+
+const HANDLE_CONVERTION_TYPE = ['string', 'number'];
+const UNIQUE_ERROR_MESSAGE = 'The date is invalid. Expected format: YYYY-MM-DD HH:mm';
+const INVALID_DATE = new Date('');
+
+function convertToDate({ schema }, value) {
+	if (value === undefined) {
```
Comment (sgendre) [inline packages/forms/src/UIForm/fields/Date/DateTime/DateTime.component.js]: what about null ?

File: packages/forms/src/UIForm/fields/Date/DateTime/DateTime.component.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+import memoizeOne from 'memoize-one';
+import InputDateTimePicker from '@talend/react-components/lib/DateTimePickers';
+import FieldTemplate from '../../FieldTemplate';
+import { isoDateTimeRegExp } from '../../../customFormats';
+import { WidgetUnhandleTypeError, WidgetUnexpectedTypeError } from '../WrongTypeError';
+import { generateDescriptionId, generateErrorId } from '../../../Message/generateId';
+
+const HANDLE_CONVERTION_TYPE = ['string', 'number'];
+const UNIQUE_ERROR_MESSAGE = 'The date is invalid. Expected format: YYYY-MM-DD HH:mm';
+const INVALID_DATE = new Date('');
```
Comment (sgendre) [inline packages/forms/src/UIForm/fields/Date/DateTime/DateTime.component.js]: not a big fan to have invalid objects, it is harder to grab the edge case from the caller, why not throw or null ?

File: packages/forms/src/UIForm/fields/Date/DateTime/DateTime.component.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+			return isoDateTimeRegExp.test(value) ? new Date(value) : INVALID_DATE;
+		default:
+			// eslint-disable-next-line no-console
+			console.error(new WidgetUnhandleTypeError(HANDLE_CONVERTION_TYPE, type));
+			return INVALID_DATE;
+	}
+}
+
+function convertFromDate({ schema }, date) {
+	if (date === undefined) {
```
Comment (sgendre) [inline packages/forms/src/UIForm/fields/Date/DateTime/DateTime.component.js]: please add null case

File: packages/forms/src/UIForm/fields/Date/DateTime/DateTime.component.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	constructor(props) {
+		super(props);
+
+		this.onChange = this.onChange.bind(this);
+		this.onBlur = this.onBlur.bind(this);
+		this.convertToDate = memoizeOne(convertToDate);
+	}
+
+	onChange(event, { datetime, errorMessage, origin }) {
+		const hasError = errorMessage !== undefined;
```
Comment (sgendre) [inline packages/forms/src/UIForm/fields/Date/DateTime/DateTime.component.js]: if errorMessage is null, it considers that there is an error?

File: packages/forms/src/UIForm/customFormats.js [hunk 5]
```diff
@@ -16,9 +16,21 @@ const urlGit = new RegExp(
...
+// EcmaScript ISO 8601 Extended Format: https://www.ecma-international.org/ecma-262/5.1/#sec-15.9.1.15
+// Only handle full string (date and time with full second fraction) and only UTC timezone
+export const isoDateTimeRegExp = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z$/;
+
 const leadingTralingSpaceRegExp = /^\s|\s$/;
 const stringWithoutSpaceRegExp = /^\S+$/;
 
+const MAX_TIMESTAMP = 8640000000000000;
+
+function isValidDate(rawString) {
```
Comment (jlame) [inline packages/forms/src/UIForm/customFormats.js]: why not directly use the date-fns function ?

File: packages/components/src/DateTimePickers/InputDateTimePicker/InputDateTimePicker.component.js [hunk 6]
```diff
@@ -33,7 +33,11 @@ const PROPS_TO_OMIT_FOR_INPUT = [
 class InputDateTimePicker extends React.Component {
 	static propTypes = {
 		id: PropTypes.string.isRequired,
-		selectedDateTime: PropTypes.instanceOf(Date),
+		selectedDateTime: PropTypes.oneOfType([
```
[the remark this answers is not in the document]
Comment (jsomsanith) [reply packages/components/src/DateTimePickers/InputDateTimePicker/InputDateTimePicker.component.js]: Done

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js [hunk 7]
```diff
@@ -380,8 +380,40 @@ function extractPartsFromTextInput(textInput, options) {
 	};
 }
 
+/**
+ * Extract parts (date, time, date/time, textInput) from a value with
+ * different possible types
+ * @param value {string | Date \ number}
```
Comment (sgendre) [inline packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js]: typo

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js [hunk 8]
```diff
@@ -380,8 +380,40 @@ function extractPartsFromTextInput(textInput, options) {
...
+ * @param options {Object}
+ * @returns
+ *	{{
+ *		date: Date,
+ *		time: { hours: string, minutes: string, seconds: string },
+ *		datetime: Date,
+ *		textInput: string
+ * 	}}
+ */
+function extractParts(value, options) {
```
Comment (sgendre) [inline packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js]: imo, this file would be better named by 'normalize.js'

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js [hunk 9]
```diff
@@ -380,8 +380,40 @@ function extractPartsFromTextInput(textInput, options) {
...
+	const typeOfValue = typeof value;
+	if (typeOfValue === 'number') {
+		return extractPartsFromDateTime(new Date(value), options);
+	} else if (typeOfValue === 'string') {
+		return extractPartsFromTextInput(value, options);
+	} else if (value instanceof Date) {
+		return extractPartsFromDateTime(value, options);
+	}
+
+	return {
```
Comment (sgendre) [inline packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js]: what about  `return null` instead of an invalid object? 
I got very bad experience with those kind of empty/invalid instances

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js [hunk 10]
```diff
@@ -43,44 +44,98 @@ describe('Date extraction', () => {
...
+			const options = { dateFormat: 'YYYY-MM-DD' };
+
+			// when
+			const parts = extractParts(validDate, options);
+
+			// then
+			expect(parts).toEqual({
+				date: new Date(2015, 8, 15),
+				datetime: validDate,
+				textInput: '2015-09-15',
```
Comment (sgendre) [inline packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js]: so is it 8 or 9?

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js [hunk 11]
```diff
@@ -43,44 +44,98 @@ describe('Date extraction', () => {
...
+			const options = { dateFormat: 'YYYY-MM-DD' };
+
+			// when
+			const parts = extractParts(timestamp, options);
+
+			// then
+			expect(parts).toEqual({
+				date: new Date(2015, 8, 15),
+				datetime: date,
+				textInput: '2015-09-15',
```
Comment (sgendre) [inline packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js]: 8 or 9? there might be something that I don't get

File: packages/forms/src/UIForm/fields/Date/README.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
+# InputDateTimePicker
+
+This widget allows you to render a date/time picker.
+
+## Data conversion
+
+You can request 3 format :
+* timestamp (number)
+* ISO-8601 ewtended format (string)
```
Comment (sgendre) [inline packages/forms/src/UIForm/fields/Date/README.md]: typo

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js (hunk 8 above)
Comment (jsomsanith) [reply packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js]: No, because we expose extraction functions. Normalize is quite generic and we usually call normalization something that takes a value, transform it and return this same value, but with a defined format.
Here we extract things : date, hours, and convert it : date and string

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js (hunk 9 above)
Comment (jsomsanith) [reply packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js]: No, the values in the returned object are used in different locations.
For example, the date and time are used in the date picker, the textInput, in the input.
All those info will live in the InputDateTimePicker state. So we need the define each one of them for setState to replace the old values.

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js (hunk 10 above)
Comment (jsomsanith) [reply packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js]: In date api 8 is the month index, that is used for september.
It is transformed to reflext `09` in the input to be human understandable.

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js (hunk 11 above)
Comment (jsomsanith) [reply packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js]: Same explanation

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js (hunk 8 above)
Comment (sgendre) [reply packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js]: inflate?

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js (hunk 10 above)
Comment (sgendre) [reply packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.test.js]: right !
it's been a while since I worked with datetime

File: packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js (hunk 8 above)
Comment (jsomsanith) [reply packages/components/src/DateTimePickers/InputDateTimePicker/date-extraction.js]: This word is not really clear neither. But I'll draw the card "Not in this PR scope" as it's not something I just created and don't want to discuss about this naming because the ideas we have for now is not better than the one in place.
But if we find something better later, an easy PR can be done.
